fsm/fsm.c

```diff
@@ -0,0 +1,206 @@
+#include <stdio.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <assert.h>
+
+typedef enum
+{
+    BOOT,
+    AUTONOMOUS,
+    MANUAL
+} State;
+
+void get_next_state(char input, State **state);
+void test_fsm();
+
+int main()
+{
+    char input;
+    State * state;
+    State tmp;
+
+    State boot = BOOT;
+
+    state = &boot;
+
+    while (1)
+    {
+        printf("\nCurrent state is: %d\n", *state);
+        printf("Please enter character input: ");
+        scanf("%c", &input);
+        if (input == '\n' || input == EOF)
+        {
+            continue;
+        }
+        if (input == 'q' || input == 'Q')
+        {
+            break;
+        }
+        get_next_state(input, &state);
+        tmp = *state;
+        fflush(stdin);
+        state = &tmp;
+    }
+
+    test_fsm();
+    printf("Unit Test Passed");
+}
+
+// Unit test asserting correct state transitions
+// returns true if pass, false otherwise
+void test_fsm()
+{
+    char input;
```

Comment:
Can you define somewhere what each char corresponds to, I figured it out after staring at the code for a while but a little key comment would be nice